libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/TripsManagerImpl.java

```diff
@@ -0,0 +1,123 @@
+  private RotateTripSku rotateSku = RotateTripSku.INVALID;
+  private AccountsPreference accountsPreference;
+
+  TripsManagerImpl(AccountsPreference accountsPreference) {
+    this.accountsPreference = accountsPreference;
+  }
+
+  private void refreshSkuToken() {
+    if (!shouldRefreshSku()) {
+      return;
+    }
+    int requestCount;
+    switch (rotateSku) {
+      case ROTATE_ON_TIMER_EXPIRE:
+        requestCount = getRouteRequestCountThreshold();
+        requestCount++;
+        setTimerExpiry();
+        persistTripsSkuToken();
+        break;
+      case ROTATE_ON_REQUEST_COUNT_EXPIRE:
+        requestCount = 0;
+        setTimerExpiry();
+        persistTripsSkuToken();
+        break;
+      default:
+        requestCount = getRouteRequestCountThreshold();
+        requestCount++;
+        break;
+    }
+    setRouteRequestCountThreshold(requestCount);
+  }
+
+  private boolean shouldRefreshSku() {
+    boolean routeReqCountExpired = validateRouteRequestCountExpiry();
+    boolean timerExpired = validateTimerExpiry();
+    if (routeReqCountExpired) {
+      rotateSku = RotateTripSku.ROTATE_ON_REQUEST_COUNT_EXPIRE;
+    } else if (timerExpired) {
+      rotateSku = RotateTripSku.ROTATE_ON_TIMER_EXPIRE;
+    } else {
+      rotateSku = RotateTripSku.INVALID;
+    }
+    return routeReqCountExpired || timerExpired;
+  }
+
+  private boolean validateTimerExpiry() {
+    long skuTokenTimeStamp = getTimerExpiry();
+    return isTwoHoursExpired(skuTokenTimeStamp);
+  }
+
+  private boolean validateRouteRequestCountExpiry() {
+    int routeRequestCount = getRouteRequestCountThreshold();
+    return routeRequestCount > ROUTE_REQUEST_COUNT_THRESHOLD;
+  }
+
+  private void setRouteRequestCountThreshold(int count) {
+    accountsPreference.getPreferenceManager().set(MAPBOX_NAV_PREFERENCE_ROUTE_REQ_COUNT, count);
+  }
+
+  private int getRouteRequestCountThreshold() {
+    return accountsPreference.getPreferenceManager().get(MAPBOX_NAV_PREFERENCE_ROUTE_REQ_COUNT, 0);
+  }
+
+  private void persistTripsSkuToken() {
+    String token = generateTripsSkuToken();
+    accountsPreference.getPreferenceManager().set(MAPBOX_NAV_PREFERENCE_TRIPS_SKU, token);
+  }
+
+  private String retrieveTripsSkuToken() {
+    return accountsPreference.getPreferenceManager().get(MAPBOX_NAV_PREFERENCE_TRIPS_SKU, "");
+  }
+
+  private String generateTripsSkuToken() {
+    return MapboxAccounts.obtainNavigationSkuSessionToken();
+  }
+
+  private void setTimerExpiry() {
+    accountsPreference.getPreferenceManager().set(MAPBOX_NAV_PREFERENCE_TRIPS_TIMESTAMP, getNow());
+  }
+
+  private long getTimerExpiry() {
+    return accountsPreference.getPreferenceManager().get(MAPBOX_NAV_PREFERENCE_TRIPS_TIMESTAMP, 0L);
+  }
+
+  private boolean isTwoHoursExpired(long then) {
+    return isExpired(getNow(), then);
+  }
+
+  private long getNow() {
+    return System.currentTimeMillis();
```

Comment:
should be careful with `System currentTimeMillis()` as this time can go back and changed by user. `SystemClock.elapsedRealTime()` is more reliable